### PR TITLE
Problem with two coexisting CKKS crypto contexts with different ring dimensions

### PR DESCRIPTION
**Pull request by dsuponitskiy**

### CODE REVIEW

**Comment by dsuponitskiy** [comment]
All CI jobs succeeded

File: src/core/lib/math/dftransform.cpp [hunk 1]
```diff
@@ -45,14 +45,29 @@
...
-size_t DiscreteFourierTransform::m_M                             = 0;
-size_t DiscreteFourierTransform::m_Nh                            = 0;
-bool DiscreteFourierTransform::m_isInitialized                   = false;
+std::unordered_map<uint32_t, DiscreteFourierTransform::PrecomputedValues> DiscreteFourierTransform::precomputedValues;
 
-/// precomputed rotation group indices
-std::vector<uint32_t> DiscreteFourierTransform::m_rotGroup;
-/// precomputed ksi powers
-std::vector<std::complex<double>> DiscreteFourierTransform::m_ksiPows;
+DiscreteFourierTransform::PrecomputedValues::PrecomputedValues(uint32_t m, uint32_t nh) {
```
Comment (yspolyakov) [inline src/core/lib/math/dftransform.cpp:50]: I would change the name of the method from PrecomputedValues to PrecomputeValues

File: src/core/lib/math/dftransform.cpp [hunk 2]
```diff
@@ -211,18 +206,19 @@ std::vector<std::complex<double>> DiscreteFourierTransform::InverseTransform(std
     return invDftRemainder;
 }
 
-void DiscreteFourierTransform::FFTSpecialInvLazy(std::vector<std::complex<double>>& vals) {
+void DiscreteFourierTransform::FFTSpecialInvLazy(std::vector<std::complex<double>>& vals,
```
Comment (yspolyakov) [inline src/core/lib/math/dftransform.cpp:209]: Why is prepValues needed? Why not read it from the static map associated with the class?

File: src/core/lib/math/dftransform.cpp (hunk 1 above)
Comment (dsuponitskiy) [reply src/core/lib/math/dftransform.cpp:50]: It is the constructor for the PrecomputedValues struct.

Comment (yspolyakov) [reply src/core/lib/math/dftransform.cpp:50]: Understood

**Review by yspolyakov — APPROVED** [review]

**Comment by yspolyakov** [comment]
I see about 2% runtime increase for `CKKSrns_Decryption` in the lib-benchmark as compared to the `dev` branch. Also, there is about 0.5$ runtime increase for `CKKSrns_Encryption`. My guess is that the extra cost is associated to resolving the reference to the STL map item. Please independently run the benchmarks to confirm my results.

**Pull request merged**